File: palindromicLinkedList/mySecondSolution.java
// link: https://leetcode.com/problems/palindrome-linked-list/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        LinkedList<ListNode> nodes = new LinkedList();
        ListNode temp = head;
        while (temp != null) {
            nodes.add(temp);
            temp = temp.next;
        }
        
        int half = nodes.size() / 2;
        temp = head;
        
        while (half > 0) {
            if (temp.val != nodes.peekLast().val)
                return false;
            temp = temp.next;
            nodes.pollLast();
            --half;
        }
        
        return true;
    }
}

File: swapNodesInPairs/my_second_solution.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def swapPairs(self, head: Optional["ListNode"]) -> Optional["ListNode"]:
        """
        head is going to be changed.
        """
        if not head or not head.next:
            return head

        new_head = None
        before = None
        start = head
        end = head.next

        while start and end:
            after = end.next

            end.next = start
            start.next = after

            new_head = end if not new_head else new_head

            if before:
                before.next = end
            before = start

            start = after
            end = None if not start else start.next

        return new_head
